File: src/demos/ps/maps/Paseo.java
package demos.ps.maps;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import static core.Script.screen;
import static demos.ps.oo.PSGame.EntFinish;
import static demos.ps.oo.PSGame.EntStart;
import static demos.ps.oo.PSGame.Shop;
import static demos.ps.oo.PSGame.getItem;
import static demos.ps.oo.PSGame.getParty;
import static demos.ps.oo.PSGame.getString;
import static demos.ps.oo.PSGame.getYesNo;
import static demos.ps.oo.PSGame.mapswitch;
import core.Script;
import demos.ps.oo.City;
import demos.ps.oo.Dungeon;
import demos.ps.oo.Enemy;
import demos.ps.oo.Item;
import demos.ps.oo.Job;
import demos.ps.oo.PSBattle;
import demos.ps.oo.PSBattle.BattleOutcome;
import demos.ps.oo.PSGame;
import demos.ps.oo.PSLibMusic.PS1Music;
import demos.ps.oo.PartyMember;
import demos.ps.oo.Specie;
import demos.ps.oo.PSGame.Flags;
import demos.ps.oo.PSGame.Planet;
import demos.ps.oo.PSLibEnemy.PS1Enemy;
import demos.ps.oo.PSLibImage.PS1Image;
import demos.ps.oo.PSLibItem.OriginalItem;
import demos.ps.oo.PSLibSound.PS1Sound;
import demos.ps.oo.PSMenu;
import demos.ps.oo.PSMenu.EntityClothes;
import demos.ps.oo.PSMenu.EntityType;
import demos.ps.oo.PSMenu.LargeEntity;
import demos.ps.oo.PSMenu.Scene;
import demos.ps.oo.PSMenu.SpecialEntity;
import demos.ps.oo.PartyMember.Gender;
import domain.VImage;

public class Paseo {

	private static final Logger log = LogManager.getLogger(Paseo.class);

	
	public static void hospital() {
		PSMenu.startScene(Scene.HOSPITAL, EntityType.CITY_WMN_BROWN, EntityClothes.WHITE);
		PSGame.Hospital(1);
		PSMenu.endScene();
	}

	public static void church() {
		PSMenu.startScene(Scene.CHURCH, SpecialEntity.PRIEST);
		PSGame.Church(1);
		PSMenu.endScene();
	}
	
	public static void weap_shop() {
		PSMenu.startScene(Scene.SHOP_WEAPON, EntityType.CITY_MAN_BROWN, EntityClothes.RED);
		Shop(getString("Shop_Weapon_Welcome"), false, new Item[]{	getItem(OriginalItem.Shield_Light_Barrier),
																	getItem(OriginalItem.Armor_Spiky_Fur),
																	getItem(OriginalItem.Armor_Zirconian_Mail)}); // changed
		PSMenu.endScene();
	}	
	
	public static void hand_shop() {
		PSMenu.startScene(Scene.SHOP_HAND, EntityType.CITY_MAN_BROWN, EntityClothes.BLUE);
		Shop(getString("Shop_Tool_Welcome"), true, new Item[]{	getItem(OriginalItem.Inventory_Flash),
																	getItem(OriginalItem.Inventory_Escape_Cloth),
																	getItem(OriginalItem.Inventory_TranCarpet)});
		PSMenu.endScene();
	}	
	
	public static void myau_shop() {
		PSMenu.startScene(Scene.SHOP_HAND, EntityType.CITY_MAN_BROWN, EntityClothes.GREEN);
		if(PSGame.hasFlag(Flags.GOT_MYAU)) {
			PSMenu.Stext(getString("Paseo_Shop_MyauAfter"));	
		} else {
			if(PSMenu.Prompt(getString("Paseo_Shop_Myau"), getYesNo()) == 1) {
				PSMenu.StextLast(getString("Paseo_Shop_MyauYes"));
			}
			else {
				if(!PSGame.getParty().hasQuestItem(getItem(OriginalItem.Quest_Laconian_Pot))) {
					PSMenu.StextLast(getString("Paseo_Shop_MyauTradeNo"));
				}
				else {
					if(PSMenu.PromptNext(getString("Paseo_Shop_MyauNo"), getYesNo()) == 1) {
						PSGame.getParty().addQuestItem(getItem(OriginalItem.Quest_Alsuline));
						PSMenu.StextLast(getString("Paseo_Shop_MyauTradeYes"));
						PSGame.getParty().removeItem(getItem(OriginalItem.Quest_Laconian_Pot));
						PSGame.setFlag(Flags.GOT_MYAU);
						
						PSGame.getParty().addMember(new PartyMember(Gender.MALE, Specie.MUSK_CAT, Job.NATURER, PSGame.getString("Name_Myau"), PS1Image.PORTRAIT_MYAU, "chars/myau.chr"));
						PSGame.getParty().getMember(1).advanceLevel();
						PSGame.getParty().getMember(1).advanceLevel();
						PSGame.getParty().getMember(1).heal();

						PSMenu.instance.back = new VImage(screen.width, screen.height); //.paintBlack();
						PSMenu.instance.entitySprite = null;
						PSGame.playMusic(PS1Music.STORY);
						screen.fadeOut(75, false);
				
						PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_ALIS), new String[]{getString("Cinematic_Myau_1")});
						PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_MYAU), new String[]{getString("Cinematic_Myau_2")});
						PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_ALIS), new String[]{getString("Cinematic_Myau_3")});
						PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_MYAU), new String[]{getString("Cinematic_Myau_4")});
						PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_ALIS), new String[]{getString("Cinematic_Myau_5")});
						PSGame.findAndPlayMusic();
						PSGame.getParty().reallocate();
					}
					else {
						PSMenu.StextLast(getString("Paseo_Shop_MyauTradeNo"));
					}
				}
			}
			
		}
		PSMenu.endScene();
	}
	
	public static void house1() {
		PSMenu.startScene(Scene.YELLOW_HOUSE, EntityType.CITY_MAN_BROWN, EntityClothes.BLUE);
		PSMenu.Stext(getString("Paseo_House_1"));
		PSMenu.endScene();
	}

	public static void house2() {
		PSMenu.startScene(Scene.YELLOW_HOUSE, EntityType.CITY_WMN_BROWN, EntityClothes.GREEN);
		PSMenu.Stext(getString("Paseo_House_2"));
		PSMenu.endScene();
	}

	public static void house3() {
		PSMenu.startScene(Scene.YELLOW_HOUSE, EntityType.CITY_MAN_BROWN, EntityClothes.GREEN);
		PSMenu.Stext(getString("Paseo_House_3"));
		PSMenu.endScene();
	}
	
	public static void ent1() {
		EntStart();
		PSMenu.Stext(getString("Paseo_People_Ent1"));
		EntFinish();
	}
	public static void ent2() {
		EntStart();
		PSMenu.Stext(getString("Paseo_People_Ent2"));
		EntFinish();
	}
	public static void ent3() {
		EntStart();
		PSMenu.Stext(getString("Paseo_People_Ent3"));
		EntFinish();
	}
	public static void ent4() {
		EntStart();
		PSMenu.Stext(getString("Paseo_People_Ent4"));
		EntFinish();
	}

	public static void tunnel_1() {
		mapswitch(Dungeon.GOVERNOR_IN);
	}
	public static void tunnel_2() {
		mapswitch(Dungeon.GOVERNOR_OUT);		
	}

	public static void rest_house() {
		PSMenu.startScene(Scene.BLUE_HOUSE, EntityType.CITY_WMN_BROWN, EntityClothes.GREEN);
		PSMenu.Stext(getString("Paseo_Governor_RestHouseBefore"));
		PSGame.playSound(PS1Sound.RESTHOUSE);
		screen.fadeOut(100, false);
		getParty().healAll(false);
		PSGame.playSound(PS1Sound.CURE);
		PSMenu.Stext(getString("Paseo_Governor_RestHouseAfter"));
		PSMenu.endScene();
	}
	
	public static void governor() {
		
		if(PSGame.hasFlag(Flags.DEFEAT_DARKFALZ)) {
			PSGame.playMusic(PS1Music.PALACE);
			PSMenu.startScene(Scene.PALACE, LargeEntity.GOVERNOR);
			PSMenu.Stext(getString("Paseo_Governor_Darkfalz"));
			if(PSMenu.PromptNext(getString("Paseo_Governor_End"), getYesNo()) == 1) {
				PSMenu.StextNext(getString("Paseo_Governor_EndYes"));	
			} else {
				PSMenu.StextNext(getString("Paseo_Governor_EndNo"));
			}
			PSMenu.StextLast(getString("Paseo_Governor_EndOk"));
			
			PSGame.endGameRoutine();
			return;
		}
		else if(PSGame.hasFlag(Flags.DEFEAT_LASSIC)) {
			PSMenu.startScene(Scene.PALACE, SpecialEntity.NONE);
			PSMenu.Stext(getString("Paseo_Mansion_Darkfalz"));
			PSGame.playSound(PS1Sound.TRAP_FALL);
			screen.fadeOut(25, false);
			PSMenu.instance.back = new VImage(screen.width, screen.height);
			PSGame.mapswitch(Dungeon.DARKFALZ_DUNGEON);
			return;
		}
		
		PSMenu.startScene(Scene.PALACE, LargeEntity.GOVERNOR);
		if(!PSGame.hasFlag(Flags.MET_GOVERNOR)) {
			PSMenu.StextNext(getString("Paseo_Governor_Intro"));
			PSGame.getParty().addQuestItem(getItem(OriginalItem.Quest_Governor_Letter));
			PSMenu.StextLast(getString("Paseo_Governor_Rest"));

			// Resting and dream
			getParty().healAll(false);
			PSGame.playSound(PS1Sound.RESTHOUSE);
			screen.fadeOut(50, false);
			VImage palace = PSMenu.instance.back; 
			PSMenu.instance.back = new VImage(screen.width, screen.height);
			PSMenu.startScene(Scene.SCREEN, SpecialEntity.NONE);
			PSMenu.StextLast(getString("Paseo_Governor_Dream"));
			PSBattle battle = new PSBattle();
			battle.startBattle(new Enemy[]{PSGame.getEnemy(PS1Enemy.SACCUBUS)}, PS1Music.BATTLE);
			getParty().healAll(true);
			PSMenu.StextLast(getString("Paseo_Governor_BadDream"));
			
			PSMenu.instance.back = palace;
			PSMenu.startScene(Scene.SCREEN, LargeEntity.GOVERNOR);	
			PSMenu.StextLast(getString("Paseo_Governor_Greet"));
			PSGame.setFlag(Flags.MET_GOVERNOR);
			
		} else {
			if(PSGame.hasFlag(Flags.GOT_NOAH)) {	
				PSMenu.StextNext(getString("Paseo_Governor_Return"));
				PSMenu.StextLast(getString("Paseo_Governor_Greet"));
			} else {
				PSMenu.StextLast(getString("Paseo_Governor_Greet_BeforeLutz"));				
			}
		}
		
		PSMenu.endScene();
	}
	
	
	public static void robot_sleep() {
		EntStart();
		PSMenu.Stext(getString("Paseo_People_Cop_Passage"));
		EntFinish();
	}
	public static void robot_exit() {
		EntStart();
		PSMenu.Stext(getString("Paseo_People_Cop_Exit"));
		EntFinish();
	}	
	
	public static void motavia() {
		mapswitch(Planet.MOTAVIA, City.PASEO.getX(),City.PASEO.getY());
	}

	public static void spaceport() {
		if(PSGame.hasFlag(Flags.GOT_HAPSBY)) {
			PSMenu.Stext(getString("Spaceport_People_Cop_Closed"));
			if(PSGame.getParty().hasQuestItem(getItem(OriginalItem.Quest_Passport))) {
				PSMenu.StextLast(getString("Spaceport_People_Cop_TakePassport"));
				PSGame.getParty().removeItem(getItem(OriginalItem.Quest_Passport));
			}
			return;
		}
		if(!PSGame.getParty().hasQuestItem(getItem(OriginalItem.Quest_Road_Pass))) {
			PSMenu.Stext(getString("Camineet_People_Cop_No_Pass"));
		} else {
			PSMenu.Stext(getString("Paseo_People_Cop_Pass"));
			mapswitch(Planet.MOTAVIA,78,35);
		}
	}

	
	
}

File: src/demos/ps/maps/Paseo.ts
/**
 * Paseo City Script
 * TypeScript port of Paseo.java
 */

import { PSGame } from '../PSGame';
import { Flags } from '../game/GameData';
import { Planet, City, CityHelper } from '../game/City';
import { Dungeon } from '../game/Dungeon';
import { OriginalItem } from '../game/PSLibItem';
import { PS1Sound } from '../game/PSLibSound';
import { PS1Music } from '../game/PSLibMusic';
import { PSSceneType, SpecialEntity, PSMenu, EntityType, EntityClothes, LargeEntity } from '../PSMenu';
import { PSMenuShop } from '../PSMenuShop';
import { PSLibEnemy, PS1Enemy } from '../game/PSLibEnemy';
import { PS1Image } from '../game/PSLibImage';
import { PSBattle } from '../battle/PSBattle';
import { PartyMember, Gender } from '../game/PartyMember';
import { Specie } from '../game/Specie';
import { Job } from '../game/Job';
import { ScriptEngine } from '../../../core/ScriptEngine';

export class Paseo {

  public static async hospital(): Promise<void> {
    await PSMenu.startScene(PSSceneType.HOSPITAL, EntityType.CITY_WMN_BROWN, EntityClothes.WHITE);
    await PSGame.Hospital(1);
    await PSMenu.endScene();
  }

  public static async church(): Promise<void> {
    await PSMenu.startScene(PSSceneType.CHURCH, SpecialEntity.PRIEST);
    await PSGame.Church(1);
    await PSMenu.endScene();
  }

  public static async weap_shop(): Promise<void> {
    await PSMenu.startScene(PSSceneType.SHOP_WEAPON, EntityType.CITY_MAN_BROWN, EntityClothes.RED);
    await PSMenuShop.Shop(PSGame.getString("Shop_Weapon_Welcome"), false, PSGame.getParty(), [
      PSGame.getItem(OriginalItem.Shield_Light_Barrier),
      PSGame.getItem(OriginalItem.Armor_Spiky_Fur),
      PSGame.getItem(OriginalItem.Armor_Zirconian_Mail)
    ]);
    await PSMenu.endScene();
  }

  public static async hand_shop(): Promise<void> {
    await PSMenu.startScene(PSSceneType.SHOP_HAND, EntityType.CITY_MAN_BROWN, EntityClothes.BLUE);
    await PSMenuShop.Shop(PSGame.getString("Shop_Tool_Welcome"), true, PSGame.getParty(), [
      PSGame.getItem(OriginalItem.Inventory_Flash),
      PSGame.getItem(OriginalItem.Inventory_Escape_Cloth),
      PSGame.getItem(OriginalItem.Inventory_TranCarpet)
    ]);
    await PSMenu.endScene();
  }

  public static async myau_shop(): Promise<void> {
    await PSMenu.startScene(PSSceneType.SHOP_HAND, EntityType.CITY_MAN_BROWN, EntityClothes.GREEN);
    if (PSGame.hasFlag(Flags.GOT_MYAU)) {
      await PSMenu.Stext(PSGame.getString("Paseo_Shop_MyauAfter"));
    } else {
      if (await PSMenu.Prompt(PSGame.getString("Paseo_Shop_Myau"), PSGame.getYesNo()) === 1) {
        await PSMenu.StextLast(PSGame.getString("Paseo_Shop_MyauYes"));
      } else {
        if (!PSGame.getParty().hasQuestItem(PSGame.getItem(OriginalItem.Quest_Laconian_Pot))) {
          await PSMenu.StextLast(PSGame.getString("Paseo_Shop_MyauTradeNo"));
        } else {
          if (await PSMenu.PromptNext(PSGame.getString("Paseo_Shop_MyauNo"), PSGame.getYesNo()) === 1) {
            PSGame.getParty().addQuestItem(PSGame.getItem(OriginalItem.Quest_Alsuline));
            await PSMenu.StextLast(PSGame.getString("Paseo_Shop_MyauTradeYes"));
            PSGame.getParty().removeItem(PSGame.getItem(OriginalItem.Quest_Laconian_Pot));
            PSGame.setFlag(Flags.GOT_MYAU);

            PSGame.getParty().addMember(new PartyMember(Gender.MALE, Specie.MUSK_CAT, Job.NATURER, PSGame.getString("Name_Myau"), PS1Image.PORTRAIT_MYAU, "Myau.anim.json"));
            PSGame.getParty().getMember(1)!.advanceLevel();
            PSGame.getParty().getMember(1)!.advanceLevel();
            PSGame.getParty().getMember(1)!.heal();

            // Java: PSMenu.instance.back = new VImage(...) - black backdrop.
            // Destroy the vendor sprite (nulling the reference alone leaks the
            // Phaser image, leaving it on screen), then fade the shop out and
            // switch to a black backdrop so the cinematic portraits draw on
            // black. A camera fade would sit on TOP of the portraits (all we'd
            // see is black), so we clear it with a fadein once the backdrop is up.
            PSMenu.instance.clearEntity();
            // A Prompt leaves its text box on the stack (see PSMenu.PromptInternal);
            // wipe those lingering vendor boxes so they don't show over the cinematic.
            PSMenu.instance.clearMenus();
            await PSGame.playMusic(PS1Music.STORY);
            await ScriptEngine.fadeout(75, false);
            PSMenu.instance.setBlackBackground();
            await ScriptEngine.fadein(1, false);

            await PSMenu.cinematicText(await PSGame.getVImage(PS1Image.CINE_ALIS), [PSGame.getString("Cinematic_Myau_1")]);
            await PSMenu.cinematicText(await PSGame.getVImage(PS1Image.CINE_MYAU), [PSGame.getString("Cinematic_Myau_2")]);
            await PSMenu.cinematicText(await PSGame.getVImage(PS1Image.CINE_ALIS), [PSGame.getString("Cinematic_Myau_3")]);
            await PSMenu.cinematicText(await PSGame.getVImage(PS1Image.CINE_MYAU), [PSGame.getString("Cinematic_Myau_4")]);
            await PSMenu.cinematicText(await PSGame.getVImage(PS1Image.CINE_ALIS), [PSGame.getString("Cinematic_Myau_5")]);
            PSGame.findAndPlayMusic();
            await PSGame.getParty().reallocate();
          } else {
            await PSMenu.StextLast(PSGame.getString("Paseo_Shop_MyauTradeNo"));
          }
        }
      }
    }
    await PSMenu.endScene();
  }

  public static async house1(): Promise<void> {
    await PSMenu.startScene(PSSceneType.YELLOW_HOUSE, EntityType.CITY_MAN_BROWN, EntityClothes.BLUE);
    await PSMenu.Stext(PSGame.getString("Paseo_House_1"));
    await PSMenu.endScene();
  }

  public static async house2(): Promise<void> {
    await PSMenu.startScene(PSSceneType.YELLOW_HOUSE, EntityType.CITY_WMN_BROWN, EntityClothes.GREEN);
    await PSMenu.Stext(PSGame.getString("Paseo_House_2"));
    await PSMenu.endScene();
  }

  public static async house3(): Promise<void> {
    await PSMenu.startScene(PSSceneType.YELLOW_HOUSE, EntityType.CITY_MAN_BROWN, EntityClothes.GREEN);
    await PSMenu.Stext(PSGame.getString("Paseo_House_3"));
    await PSMenu.endScene();
  }

  public static async ent1(): Promise<void> {
    PSGame.EntStart();
    await PSMenu.Stext(PSGame.getString("Paseo_People_Ent1"));
    PSGame.EntFinish();
  }

  public static async ent2(): Promise<void> {
    PSGame.EntStart();
    await PSMenu.Stext(PSGame.getString("Paseo_People_Ent2"));
    PSGame.EntFinish();
  }

  public static async ent3(): Promise<void> {
    PSGame.EntStart();
    await PSMenu.Stext(PSGame.getString("Paseo_People_Ent3"));
    PSGame.EntFinish();
  }

  public static async ent4(): Promise<void> {
    PSGame.EntStart();
    await PSMenu.Stext(PSGame.getString("Paseo_People_Ent4"));
    PSGame.EntFinish();
  }

  public static async tunnel_1(): Promise<void> {
    await PSGame.mapswitchToDungeon(Dungeon.GOVERNOR_IN);
  }

  public static async tunnel_2(): Promise<void> {
    await PSGame.mapswitchToDungeon(Dungeon.GOVERNOR_OUT);
  }

  public static async rest_house(): Promise<void> {
    await PSMenu.startScene(PSSceneType.BLUE_HOUSE, EntityType.CITY_WMN_BROWN, EntityClothes.GREEN);
    await PSMenu.Stext(PSGame.getString("Paseo_Governor_RestHouseBefore"));
    PSGame.playSound(PS1Sound.RESTHOUSE);
    await ScriptEngine.fadeout(100, false);
    PSGame.getParty().healAll(false);
    // Rest a while on the black screen before waking up (the Java fade loop
    // provided this pause implicitly by blocking for its full duration)
    await PSMenu.instance.waitDelay(125);
    PSGame.playSound(PS1Sound.CURE);
    // Java's fadeOut is transient — the next render pops the screen back. The
    // Phaser camera fade persists until an explicit fadein, so without this
    // the closing text (and the rest of the scene) plays on a black screen.
    await ScriptEngine.fadein(1, false);
    await PSMenu.Stext(PSGame.getString("Paseo_Governor_RestHouseAfter"));
    await PSMenu.endScene();
  }

  public static async governor(): Promise<void> {
    if (PSGame.hasFlag(Flags.DEFEAT_DARKFALZ)) {
      await PSGame.playMusic(PS1Music.PALACE);
      await PSMenu.startSceneWithLargeEntity(PSSceneType.PALACE, LargeEntity.GOVERNOR);
      await PSMenu.Stext(PSGame.getString("Paseo_Governor_Darkfalz"));
      if (await PSMenu.PromptNext(PSGame.getString("Paseo_Governor_End"), PSGame.getYesNo()) === 1) {
        await PSMenu.StextNext(PSGame.getString("Paseo_Governor_EndYes"));
      } else {
        await PSMenu.StextNext(PSGame.getString("Paseo_Governor_EndNo"));
      }
      await PSMenu.StextLast(PSGame.getString("Paseo_Governor_EndOk"));

      await PSGame.endGameRoutine();
      return;
    } else if (PSGame.hasFlag(Flags.DEFEAT_LASSIC)) {
      await PSMenu.startScene(PSSceneType.PALACE, SpecialEntity.NONE);
      await PSMenu.Stext(PSGame.getString("Paseo_Mansion_Darkfalz"));
      PSGame.playSound(PS1Sound.TRAP_FALL);
      await ScriptEngine.fadeout(25, false);
      // Java: PSMenu.instance.back = new VImage(...) - black backdrop

      await PSGame.mapswitchToDungeon(Dungeon.DARKFALZ_DUNGEON);
      return;
    }

    await PSMenu.startSceneWithLargeEntity(PSSceneType.PALACE, LargeEntity.GOVERNOR);
    if (!PSGame.hasFlag(Flags.MET_GOVERNOR)) {
      await PSMenu.StextNext(PSGame.getString("Paseo_Governor_Intro"));
      PSGame.getParty().addQuestItem(PSGame.getItem(OriginalItem.Quest_Governor_Letter));
      await PSMenu.StextLast(PSGame.getString("Paseo_Governor_Rest"));

      // Resting and dream
      PSGame.getParty().healAll(false);
      PSGame.playSound(PS1Sound.RESTHOUSE);
      await ScriptEngine.fadeout(50, false);

      // Java: back = new VImage(...) — the dream plays on a black backdrop
      // (SCREEN scenes keep whatever backdrop is set). Depth 1950 keeps the
      // black at background level so the battle enemy sprites (1994) show.
      PSMenu.instance.setBlackBackground(1950);

      await PSMenu.startScene(PSSceneType.SCREEN, SpecialEntity.NONE);
      await PSMenu.StextLast(PSGame.getString("Paseo_Governor_Dream"));

      const battle = new PSBattle();
      await battle.startBattle([PSLibEnemy.getEnemyByEnum(PS1Enemy.SACCUBUS)!], PS1Music.BATTLE);

      PSGame.getParty().healAll(true);
      await PSMenu.StextLast(PSGame.getString("Paseo_Governor_BadDream"));

      // Java: back = palace — waking up restores the palace backdrop
      PSMenu.instance.setBackground(PSGame.getImage(PSSceneType.PALACE));
      await PSMenu.startSceneWithLargeEntity(PSSceneType.SCREEN, LargeEntity.GOVERNOR);
      await PSMenu.StextLast(PSGame.getString("Paseo_Governor_Greet"));
      PSGame.setFlag(Flags.MET_GOVERNOR);
    } else {
      if (PSGame.hasFlag(Flags.GOT_NOAH)) {
        await PSMenu.StextNext(PSGame.getString("Paseo_Governor_Return"));
        await PSMenu.StextLast(PSGame.getString("Paseo_Governor_Greet"));
      } else {
        await PSMenu.StextLast(PSGame.getString("Paseo_Governor_Greet_BeforeLutz"));
      }
    }

    await PSMenu.endScene();
  }

  public static async robot_sleep(): Promise<void> {
    PSGame.EntStart();
    await PSMenu.Stext(PSGame.getString("Paseo_People_Cop_Passage"));
    PSGame.EntFinish();
  }

  public static async robot_exit(): Promise<void> {
    PSGame.EntStart();
    await PSMenu.Stext(PSGame.getString("Paseo_People_Cop_Exit"));
    PSGame.EntFinish();
  }

  public static motavia(): void {
    PSGame.mapswitchToPlanet(Planet.MOTAVIA, CityHelper.getX(City.PASEO), CityHelper.getY(City.PASEO));
  }

  public static async spaceport(): Promise<void> {
    PSGame.EntStart();
    if (PSGame.hasFlag(Flags.GOT_HAPSBY)) {
      await PSMenu.Stext(PSGame.getString("Spaceport_People_Cop_Closed"));
      if (PSGame.getParty().hasQuestItem(PSGame.getItem(OriginalItem.Quest_Passport))) {
        await PSMenu.StextLast(PSGame.getString("Spaceport_People_Cop_TakePassport"));
        PSGame.getParty().removeItem(PSGame.getItem(OriginalItem.Quest_Passport));
      }
      PSGame.EntFinish();
      return;
    }
    if (!PSGame.getParty().hasQuestItem(PSGame.getItem(OriginalItem.Quest_Road_Pass))) {
      await PSMenu.Stext(PSGame.getString("Camineet_People_Cop_No_Pass"));
    } else {
      await PSMenu.Stext(PSGame.getString("Paseo_People_Cop_Pass"));
      PSGame.mapswitchToPlanet(Planet.MOTAVIA, 78, 35);
    }
    PSGame.EntFinish();
  }
}